ProjFS.Mac/PrjFSKext/PrjFSKext/AssociativeArray.hpp

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include "Locks.hpp"
+#include "Memory.hpp"
+#include "KextCppUtilities.hpp"
+
+struct ResizeResult
+{
+    bool   memoryWasUsed;
+    uint32_t releasedMemorySize;
+    void*  releasedMemory;
+};
+
+template <typename KEY, typename VALUE, bool EQUAL_FN(const KEY&, const VALUE&)>
+class AssociativeArray
+{
+    AssociativeArray(const AssociativeArray&) = delete;
+    AssociativeArray& operator=(const AssociativeArray&) = delete;
+
+    uint32_t capacity;
+    uint32_t numberUsed;
+    VALUE* data;
+    
+    uint32_t NextGrowCapacity() const
+    {
+        return (this->capacity == 0) ? 8u : (this->capacity * 2u);
+    }
+    
+public:
+    AssociativeArray() :
+        capacity(0),
+        numberUsed(0),
+        data(nullptr)
+    {}
+    
+    ~AssociativeArray()
+    {
+        this->ClearAndFreeMemory();
+    }
+    
+    void Clear()
+    {
+        for (uint32_t i = 0; i < this->numberUsed; ++i)
+        {
+            this->data[i].~VALUE();
+        }
+        this->numberUsed = 0;
+    }
+    
+    void ClearAndFreeMemory()
+    {
+        this->Clear();
+        
+        if (nullptr != this->data)
+        {
+            Memory_Free(this->data, MemorySizeForCapacity(this->capacity));
+            this->data = nullptr;
+            this->capacity = 0;
+        }
+    }
+    
+    uint32_t MemorySizeToGrow() const
+    {
+        return this->NextGrowCapacity() * sizeof(this->data[0]);
+    }
+    
+    static uint32_t MemorySizeForCapacity(uint32_t capacity)
+    {
+        return capacity * sizeof(VALUE);
+    }
+    
+    ResizeResult GrowUsingMemory(void* memory, uint32_t memorySize)
```

Comment:
Does the deadlock avoidance/lock-dropping explanation clarify this aspect of the API design? Any alternative suggestions bearing in mind that constraint? (I deliberately decomposed it so that you can either call the kitchen sink find-insert-resize-possibly-dropping-lock function, or you can call these individual functions in situations where momentarily dropping the lock is not so simple.)